method/publish.json

```diff
@@ -0,0 +1,56 @@
+{
+	"$schema": "http://json-schema.org/draft/2019-09/schema#",
+	"$id": "http://pop2020.dedis.ch/publish.json",
+	"description": "Match publish query",
+	"type": "object",
+
+	"properties": {
+		"method": { 
+			"description": "[String] operation to be performed by the query",
+			"const": "publish" 
+		},
+		"params": {
+			"oneOf": [
+				{
+					"description": "Match a publish query destined for the root channel",
+					"type": "object",
+					"properties": {
+						"channel": { 
+							"description": "[String] name of the (root) channel", 
+							"$ref": "channel/rootChannel.json",
+							"$comment": "Note: should always be \"/root\""
+						},
+						"message": { 
+							"description": "[Message] message to be published",
+							"$ref": "message/messageGeneral.json"
+						}
+					},
+					"additionalProperties": false,
+					"required": ["channel", "message"]
+				}, 
+				{
+					"description": "Match a publish query destined for a sub-channel",
+					"type": "object",
+					"properties": {
+						"channel": { 
+							"description": "[String] name of the channel", 
+							"$ref": "channel/subChannel.json"
+						},

```

Comment:
The only difference here is in the channel.
You might want to match messageGeneral and just have a one-of for the channel specification.
This can save you a lot of repetition.

...then the question is: do you really want to verify the channel, or just match any channel? (at this level)

Keep in mind, JSON-Schema is a specification, but you can (and absolutely should) do application-level verification, beyond what JSON-Schema could provide anyways.

Reply:
In this piece of code, I think we have to separate both parts because there are *two* differences:

- channel name (`/root` or `/root/<channel>`);
- message type (a "witness a message" message cannot be published on the main channel)

I think there's a way to do it using if-then-else